QS: Dynamic Metadata with Automation (#1464)
smartcontractkit/documentation: src/content/chainlink-automation/tutorials/dynamic-nft-qs.mdx

---
section: automation
date: Last Modified

> title: Creating Dynamic NFTs"

Reviewer: ```suggestion
title: Creating Dynamic NFTs
```

Learn how to create NFTs that can change over time, based on user interaction, or any number of conditions by following along with this video tutorial:

<YouTube id="https://www.youtube.com/watch?v=E7Rm1LUKhj4" />

{/* prettier-ignore */}
<div class="remix-callout">
  <a href="https://github.com/smartcontractkit/smart-contract-examples/tree/main/dynamic-nft">See the code on GitHub</a>
</div>

<Aside type="caution" title="Disclaimer">
  This tutorial represents an example of using a Chainlink product or service and is provided to help you understand how
  to interact with Chainlink’s systems and services so that you can integrate them into your own. This template is
  provided “AS IS” and “AS AVAILABLE” without warranties of any kind, has not been audited, and may be missing key
  checks or error handling to make the usage of the product more clear. Do not use the code in this example in a
  production environment without completing your own audits and application of best practices. Neither Chainlink Labs,
  the Chainlink Foundation, nor Chainlink node operators are responsible for unintended outputs that are generated due
  to errors in code.
</Aside>

> ## Overview

Reviewer 2: Let's use the default Overview right-nav section with no heading in the content for the intro paragraph. Where applicable, add a specific "Objective" section. This content will eventually go into the new QuickStarts page format as shown in https://github.com/smartcontractkit/documentation/pull/1458.

I was originally planning to skip the Objective sections but I think we can make them work. For this example, just rename `## Overview` to `## Objective`.

> ## Overview

Author: I'm renaming this section to Objective. Thanks.

In this tutorial, you will use the [Chainlink Automation Network](https://automation.chain.link) to help create a dynamic NFT by using the Automation-compatible smart contract provided in the repository.

The demo includes:

- Setting up and funding an upkeep on the Chainlink Automation Network
- Setting up a dynamic NFT and viewing the progression of the NFT on [OpenSea](https://testnets.opensea.io/).

## Before you begin

Before you start this tutorial, complete the following items:

- Set up a cryptocurrency wallet
- To deploy this contract on testnets, ensure the deployer account has testnet LINK. Use the [LINK faucet](https://faucets.chain.link/) to retrieve testnet LINK.
- Open a new [Remix](https://remix.ethereum.org/) project.

## Run steps

> 1. Open the `2_complete.sol` contract in Remix and rename the contract to `dynNFT.sol`.

Reviewer: <img width="231" alt="Screen Shot 2023-08-24 at 4 13 06 PM" src="https://github.com/smartcontractkit/documentation/assets/5498502/42789a9b-d8ac-4b2b-a179-3d9f85756d51">

This CSS issue has come up on a few of my PRs. Two possible solutions:

- Works sometimes, a bit finicky: Make sure this bit of HTML is all in one line, using prettier-ignore if necessary
- Works more reliably: Use the `CodeSample` component with `showButtonOnly` parameter

Author: CodeSample works only w our local files; checked over this. Resolved for now. may use prettier-ignore

   <div class="remix-callout">
     <a href="https://github.com/smartcontractkit/smart-contract-examples/blob/main/dynamic-nft/2_complete.sol">
       Open in Remix
     </a>
   </div>

1. Compile the contract in Remix. Under the section titles **DEPLOY & RUN TRANSACTIONS** you will have to change some values in the dropdown menus. Under **ENVIRONMENT**, select _Injected Provider - MetaMask_ or _WalletConnect_ depending on the cryptocurrency wallet you are using. Selecting one of these options will prompt you to connect your wallet with Remix.

   <ClickToZoom src="/images/automation/dynnft-qs-1.png" alt="Dynamic NFT Quickstart Step 2" style="max-width: 70%;" />

1. Ensure that the **CONTRACT** field displays `dynNFT`. Finally, for the **Deploy** value, you must enter a time interval value denoting when the dynamic NFT will change. For this demo, we have inputted 30, but you can change this value. Once you have entered your interval, click the orange **transact** button and follow the prompts to allow the transaction to be performed. This may take a few minutes depending on factors such as network traffic.

> 1. Once the contract is deployed, you should be able to see more information under the **Deployed Contracts** section. You must mint an NFT. In the field next to **safeMint** paste your wallet address. Then, click the orange button and follow the prompts to comfirm the transaction.

Reviewer: ```suggestion
1. Once the contract is deployed, you should be able to see more information under the **Deployed Contracts** section. 

   At this point, you can mint an NFT. In the field next to **safeMint** paste your wallet address. Then, click the orange **transact** button and follow the prompts to confirm the transaction.
```

   <ClickToZoom src="/images/automation/dynnft-qs-2.png" alt="Dynamic NFT Quickstart Step 4" style="max-width: 70%;" />

> 1. After minting the NFT, you can look at the stage of the NFT based on the tokenURIs given in the contract. Enter `0` in the **flowerStage** field and click the blue button. You should see the stage of the NFT. Enter `0` in the **tokenURI** field and click the blue button to see the corresponding URI for the NFT.

Reviewer: ```suggestion
1. After minting the NFT, you can look at the stage of the NFT based on the tokenURIs given in the contract. Enter `"0"` in the **flowerStage** field and click the blue **call** button. You should see the stage of the NFT. 

   Enter `0` in the **tokenURI** field and click the blue **call** button to see the corresponding URI for the NFT.
```

   <ClickToZoom
     src="/images/automation/dynnft-qs-3.png"
     alt="Dynamic NFT Quickstart Step 5.1"
     style="max-width: 70%;"
   />
   <ClickToZoom
     src="/images/automation/dynnft-qs-4.png"
     alt="Dynamic NFT Quickstart Step 5.2"
     style="max-width: 70%;"
   />

1. Copy the contract address and open the [Chainlink Automation Interface](https://automation.chain.link). Connect your wallet to the network and click the blue button **Register New Upkeep**. For the Trigger mechanism, select Custom logic. You will prompted to enter your contract address. Paste your address and click **Next**.

   <ClickToZoom src="/images/automation/dynnft-qs-5.png" alt="Dynamic NFT Quickstart Step 6" style="max-width: 70%;" />

1. You must fill out the prompts to finish registering upkeep. For the **Starting balance (LINK)**, enter 0.5 for this demo. After entering information, click **Register Upkeep**. It may take a few moments for this to complete.

> 1. Navigate back to Remix. Under the **Deployed Contracts** section, enter `[]` in the **checkUpkeep** field and click the blue button. You may see the following text: `upkeepNeeded true`. This means that an upkeep will be performed.

Author: Grabbed an image; thanks. Resolved

  to interact with Chainlink’s systems and services so that you can integrate them into your own. This template is
  provided “AS IS” and “AS AVAILABLE” without warranties of any kind, has not been audited, and may be missing key
  checks or error handling to make the usage of the product more clear. Do not use the code in this example in a
  production environment without completing your own audits and application of best practices. Neither Chainlink Labs,
  the Chainlink Foundation, nor Chainlink node operators are responsible for unintended outputs that are generated due
  to errors in code.
</Aside>

## Overview

In this tutorial, you will use the [Chainlink Automation Network](https://automation.chain.link) to help create a dynamic NFT by using the Automation-compatible smart contract provided in the repository.

The demo includes:

- Setting up and funding an upkeep on the Chainlink Automation Network
- Setting up a dynamic NFT and viewing the progression of the NFT on [OpenSea](https://testnets.opensea.io/).

## Before you begin

Before you start this tutorial, complete the following items:

> - Set up a cryptocurrency wallet
> - To deploy this contract on testnets, ensure the deployer account has testnet LINK. Use the [LINK faucet](https://faucets.chain.link/) to retrieve testnet LINK.
> - Open a new [Remix](https://remix.ethereum.org/) project.

Reviewer 2: ```suggestion
- If you are new to smart contract development, learn how to [Deploy Your First Smart Contract](/getting-started/deploy-your-first-contract).
- Set up a cryptocurrency wallet
- To deploy this contract on testnets, ensure the deployer account has testnet LINK. Use the [LINK faucet](https://faucets.chain.link/) to retrieve testnet LINK.
- Open a new [Remix](https://remix.ethereum.org/) project.
```

- Set up a cryptocurrency wallet
- To deploy this contract on testnets, ensure the deployer account has testnet LINK. Use the [LINK faucet](https://faucets.chain.link/) to retrieve testnet LINK.
- Open a new [Remix](https://remix.ethereum.org/) project.

## Run steps

1. Open the `2_complete.sol` contract in Remix and rename the contract to `dynNFT.sol`.

   {/* prettier-ignore */}
   <div class="remix-callout">
     <a href="https://remix.ethereum.org/#url=https://github.com/smartcontractkit/smart-contract-examples/blob/main/dynamic-nft/2_complete.sol">Open in Remix</a>
   </div>

1. Compile the contract in Remix. Under the section titles **DEPLOY & RUN TRANSACTIONS** you will have to change some values in the dropdown menus. Under **ENVIRONMENT**, select _Injected Provider - MetaMask_ or _WalletConnect_ depending on the cryptocurrency wallet you are using. Selecting one of these options will prompt you to connect your wallet with Remix.

   <ClickToZoom src="/images/automation/dynnft-qs-1.png" alt="Dynamic NFT Quickstart Step 2" style="max-width: 70%;" />

> 1. Ensure that the **CONTRACT** field displays `dynNFT`. Finally, for the **Deploy** value, you must enter a time interval value denoting when the dynamic NFT will change. For this demo, we have inputted 30, but you can change this value. Once you have entered your interval, click the orange **transact** button and follow the prompts to allow the transaction to be performed. This may take a few minutes depending on factors such as network traffic.

Author: Went with the former, thank you. Resolved

  checks or error handling to make the usage of the product more clear. Do not use the code in this example in a
  production environment without completing your own audits and application of best practices. Neither Chainlink Labs,
  the Chainlink Foundation, nor Chainlink node operators are responsible for unintended outputs that are generated due
  to errors in code.
</Aside>

## Overview

In this tutorial, you will use the [Chainlink Automation Network](https://automation.chain.link) to help create a dynamic NFT by using the Automation-compatible smart contract provided in the repository.

The demo includes:

- Setting up and funding an upkeep on the Chainlink Automation Network
- Setting up a dynamic NFT and viewing the progression of the NFT on [OpenSea](https://testnets.opensea.io/).

## Before you begin

Before you start this tutorial, complete the following items:

- Set up a cryptocurrency wallet
- To deploy this contract on testnets, ensure the deployer account has testnet LINK. Use the [LINK faucet](https://faucets.chain.link/) to retrieve testnet LINK.
- Open a new [Remix](https://remix.ethereum.org/) project.

## Run steps

> 1. From the [GitHub repository](https://github.com/smartcontractkit/smart-contract-examples/tree/main/dynamic-nft) copy the code from [`2-complete.sol`](https://github.com/smartcontractkit/smart-contract-examples/blob/main/dynamic-nft/2_complete.sol). Create a new file in Remix and name it `dynNFT.sol`. Paste the code into this file.

Reviewer 2: Can this be an Open in Remix button?

Author: Added button, resolved

1. While waiting for the upkeep to be performed, copy the contract address to your clipboard and open OpenSea(https://testnets.opensea.io/). Paste your address in search bar at the top of the page. You should be directed to a page with the image of your minted NFT. Click the seed thumbnail to learn more about the NFT.

1. Navigate back to Remix. Enter `[]` in the **checkUpkeep** field and click the blue button. If this returns `upkeepNeeded false` your upkeep should have been performed. Navigate to the [Chainlink Automation Interface](automation.chain.link) and select your upkeep to view more details and to confirm that an upkeep was performed. If `upkeepNeeded false` is not returned in Remix, you may have to wait a few minutes and run the **checkUpkeep** function.

1. Once the upkeep has been performed, navigate to Remix and click **flowerStage** and **tokenURI** to see the most recent flower stage and corresponding URI. Navigate back to the information of your NFT on OpenSea. Click the button on the top right to refresh the metadata. You should see the NFT become a bloom.

> 1. To see the NFT become a fully blossomed flower, you will have to wait another 30 seconds (or the amount of time you denoted when deploying your contract). After this interval, refresh the NFT's metadata again. You should see a fully blossomed flower. You have now completed the tutorial.

Reviewer 2: Add a summary of what this quickstart accomplished.

Author: Added summary; resolved
